File: workers/ts-adapter/src/function-candidates.ts
import ts from "typescript";

import { createEmptyBoundaryImpact, createEmptyContextSignals } from "./protocol.ts";
import type { CandidatePayload } from "./protocol.ts";
import { lineSpan, relativePosix } from "./project.ts";
import {
  collectTopLevelFunctionLikes,
  countSymbolReferences,
  duplicateFunctionKey,
  isExportedNode,
  passthroughTarget,
} from "./syntax.ts";

const LONG_FUNCTION_THRESHOLD = 40;
const DUPLICATE_FUNCTION_MIN_LINES = 3;
const INLINE_FUNCTION_MAX_LINES = 8;

export function buildLongFunctionCandidates(sourceFile: ts.SourceFile, root: string): CandidatePayload[] {
  const relPath = relativePosix(root, sourceFile.fileName);
  const totalLines = sourceFile.getLineAndCharacterOfPosition(sourceFile.end).line + 1;
  const candidates: CandidatePayload[] = [];

  for (const entry of collectTopLevelFunctionLikes(sourceFile)) {
    const span = lineSpan(sourceFile, entry.declaration);
    if (span.length < LONG_FUNCTION_THRESHOLD) {
      continue;
    }
    candidates.push({
      id: `ts-extract-function-${relPath}-${span.startLine}-${entry.nameNode.text}`,
      kind: "extract_function",
      title: `Extract logic from long function ${entry.nameNode.text}`,
      description: `Function \`${entry.nameNode.text}\` in ${relPath} spans ${span.length} lines and is a candidate for extraction`,
      language: "typescript",
      scope: "local",
      source: ["static", "metric"],
      files: [relPath],
      symbols: [entry.nameNode.text],
      anchorRegions: [{ file: relPath, startLine: span.startLine, endLine: span.endLine }],
      estimatedBenefit: {
        complexityReduction: Math.min(1, span.length / Math.max(totalLines, 1)),
        maintainabilityGain: 0.35,
      },
      estimatedRisk: { semanticRisk: 0.35, testRisk: 0.25, conflictRisk: 0.15 },
      estimatedDiff: {
        filesTouched: 1,
        linesAdded: Math.max(3, Math.floor(span.length / 4)),
        linesModified: span.length,
      },
      contextSignals: createEmptyContextSignals(),
      boundaryImpact: createEmptyBoundaryImpact(),
      confidence: 0.68,
      applyModeHint: "guarded",
      requiredChecks: ["parse", "lint", "typecheck", "unit_test"],
      dependencies: [],
      conflicts: [],
      provenance: {
        detectors: ["ts-worker-long-function"],
        evidence: [`line_span:${span.length}`, `symbol:${entry.nameNode.text}`],
      },
    });
  }

  return candidates;
}

export function buildDuplicateFunctionCandidates(sourceFile: ts.SourceFile, root: string): CandidatePayload[] {
  const relPath = relativePosix(root, sourceFile.fileName);
  const groups = new Map<string, Array<{ name: string; startLine: number; endLine: number; length: number }>>();

  for (const entry of collectTopLevelFunctionLikes(sourceFile)) {
    const duplicateKey = duplicateFunctionKey(sourceFile, entry);
    if (!duplicateKey) {
      continue;
    }
    const span = lineSpan(sourceFile, entry.declaration);
    const groupEntry = {
      name: entry.nameNode.text,
      startLine: span.startLine,
      endLine: span.endLine,
      length: span.length,
    };
    groups.set(duplicateKey, [...(groups.get(duplicateKey) ?? []), groupEntry]);
  }

  const candidates: CandidatePayload[] = [];
  for (const functions of [...groups.values()].sort((left, right) => left[0]!.startLine - right[0]!.startLine)) {
    if (functions.length < 2) {
      continue;
    }
    if (Math.max(...functions.map((entry) => entry.length)) < DUPLICATE_FUNCTION_MIN_LINES) {
      continue;
    }
    const symbols = functions.map((entry) => entry.name);
    const totalLines = functions.reduce((sum, entry) => sum + entry.length, 0);
    candidates.push({
      id: `ts-duplicate-logic-${relPath}-${functions[0]!.startLine}-${functions.length}`,
      kind: "duplicate_logic",
      title: `Consolidate duplicate TypeScript functions in ${relPath}`,
      description: `Functions ${symbols.map((symbol) => `\`${symbol}\``).join(", ")} in ${relPath} share the same structure and are candidates for consolidation`,
      language: "typescript",
      scope: "module",
      source: ["clone", "metric"],
      files: [relPath],
      symbols,
      anchorRegions: functions.map((entry) => ({ file: relPath, startLine: entry.startLine, endLine: entry.endLine })),
      estimatedBenefit: { duplicationReduction: Math.min(1, functions.length / 3), maintainabilityGain: 0.42 },
      estimatedRisk: { semanticRisk: 0.28, apiRisk: 0.12, testRisk: 0.22, conflictRisk: 0.18 },
      estimatedDiff: {
        filesTouched: 1,
        linesAdded: Math.max(3, Math.floor(totalLines / 6)),
        linesModified: totalLines,
      },
      contextSignals: createEmptyContextSignals(),
      boundaryImpact: createEmptyBoundaryImpact(),
      confidence: 0.74,
      applyModeHint: "guarded",
      requiredChecks: ["parse", "lint", "typecheck", "unit_test"],
      dependencies: [],
      conflicts: [],
      provenance: {
        detectors: ["ts-worker-duplicate-function"],
        evidence: symbols.map((symbol) => `symbol:${symbol}`).concat(`duplicateGroupSize:${functions.length}`),
      },
    });
  }

  return candidates;
}

export function buildRemoveAbstractionCandidates(sourceFile: ts.SourceFile, root: string): CandidatePayload[] {
  const relPath = relativePosix(root, sourceFile.fileName);
  const candidates: CandidatePayload[] = [];

  for (const entry of collectTopLevelFunctionLikes(sourceFile)) {
    if (!entry.nameNode.text.startsWith("_")) {
      continue;
    }
    const target = passthroughTarget(entry);
    if (!target) {
      continue;
    }
    const span = lineSpan(sourceFile, entry.declaration);
    candidates.push({
      id: `ts-remove-abstraction-${relPath}-${span.startLine}-${entry.nameNode.text}`,
      kind: "remove_abstraction",
      title: `Inline thin wrapper ${entry.nameNode.text}`,
      description: `Private TypeScript wrapper \`${entry.nameNode.text}\` in ${relPath} only forwards to \`${target}\` and can likely be removed`,
      language: "typescript",
      scope: "module",
      source: ["static", "metric"],
      files: [relPath],
      symbols: [entry.nameNode.text],
      anchorRegions: [{ file: relPath, startLine: span.startLine, endLine: span.endLine }],
      estimatedBenefit: {
        complexityReduction: Math.min(1, span.length / Math.max(LONG_FUNCTION_THRESHOLD, 1)),
        maintainabilityGain: 0.26,
      },
      estimatedRisk: { semanticRisk: 0.2, apiRisk: 0.08, testRisk: 0.18, conflictRisk: 0.12 },
      estimatedDiff: {
        filesTouched: 1,
        linesAdded: Math.max(1, Math.floor(span.length / 3)),
        linesModified: span.length,
      },
      contextSignals: createEmptyContextSignals(),
      boundaryImpact: createEmptyBoundaryImpact(),
      confidence: 0.74,
      applyModeHint: "guarded",
      requiredChecks: ["parse", "lint", "typecheck", "unit_test"],
      dependencies: [],
      conflicts: [],
      provenance: {
        detectors: ["ts-worker-passthrough-wrapper"],
        evidence: [`symbol:${entry.nameNode.text}`, `target:${target}`, `line_span:${span.length}`],
      },
    });
  }

  return candidates;
}

export function buildInlineFunctionCandidates(
  checker: ts.TypeChecker,
  sourceFile: ts.SourceFile,
  root: string,
): CandidatePayload[] {
  const relPath = relativePosix(root, sourceFile.fileName);
  const usageCounts = countSymbolReferences(checker, sourceFile);
  const candidates: CandidatePayload[] = [];

  for (const entry of collectTopLevelFunctionLikes(sourceFile)) {
    if (!entry.nameNode.text.startsWith("_") || isExportedNode(entry.declaration)) {
      continue;
    }
    if (passthroughTarget(entry)) {
      continue;
    }
    const symbol = checker.getSymbolAtLocation(entry.nameNode);
    if (!symbol || (usageCounts.get(symbol) ?? 0) !== 1) {
      continue;
    }
    const span = lineSpan(sourceFile, entry.declaration);
    if (span.length > INLINE_FUNCTION_MAX_LINES) {
      continue;
    }
    candidates.push({
      id: `ts-inline-function-${relPath}-${span.startLine}-${entry.nameNode.text}`,
      kind: "inline_function",
      title: `Inline single-use helper ${entry.nameNode.text}`,
      description: `Private TypeScript helper \`${entry.nameNode.text}\` in ${relPath} is referenced only once and is a candidate for inlining into its caller`,
      language: "typescript",
      scope: "module",
      source: ["static", "metric"],
      files: [relPath],
      symbols: [entry.nameNode.text],
      anchorRegions: [{ file: relPath, startLine: span.startLine, endLine: span.endLine }],
      estimatedBenefit: {
        complexityReduction: Math.min(1, span.length / Math.max(INLINE_FUNCTION_MAX_LINES, 1)),
        maintainabilityGain: 0.24,
      },
      estimatedRisk: { semanticRisk: 0.24, apiRisk: 0.06, testRisk: 0.18, conflictRisk: 0.1 },
      estimatedDiff: {
        filesTouched: 1,
        linesAdded: Math.max(1, Math.floor(span.length / 2)),
        linesModified: span.length,
      },
      contextSignals: createEmptyContextSignals(),
      boundaryImpact: createEmptyBoundaryImpact(),
      confidence: 0.72,
      applyModeHint: "guarded",
      requiredChecks: ["parse", "lint", "typecheck", "unit_test"],
      dependencies: [],
      conflicts: [],
      provenance: {
        detectors: ["ts-worker-single-use-helper"],
        evidence: [`symbol:${entry.nameNode.text}`, "referenceCount:1", `line_span:${span.length}`],
      },
    });
  }

  return candidates;
}

File: refactorq/adapters/python/function_candidates.py
from __future__ import annotations

import ast

from refactorq.adapters.python.builders import top_level_unused_assignment_candidate
from refactorq.adapters.python.common import (
    INLINE_FUNCTION_MAX_LINES,
    LONG_FUNCTION_THRESHOLD,
    benefit,
    diff,
    duplicate_function_key,
    exported_names as collect_exported_names,
    loaded_name_counts as collect_loaded_name_counts,
    passthrough_target,
    region,
    risk,
)
from refactorq.core.candidate.models import Candidate, Provenance


def collect_name_context(tree: ast.Module) -> tuple[dict[str, int], set[str], set[str]]:
    loaded_name_counts = collect_loaded_name_counts(tree)
    return loaded_name_counts, set(loaded_name_counts), collect_exported_names(tree)


def collect_function_candidates(
    rel_path: str,
    lines: list[str],
    loaded_name_counts: dict[str, int],
    exported_names: set[str],
    candidates: list[Candidate],
    duplicate_functions: list[tuple[str, int, int, str]],
    passthrough_functions: list[tuple[str, int, int, str]],
    inline_functions: list[tuple[str, int, int, int]],
    node: ast.FunctionDef | ast.AsyncFunctionDef,
) -> None:
    end_line = node.end_lineno
    assert end_line is not None
    length = end_line - node.lineno + 1
    duplicate_functions.append((node.name, node.lineno, end_line, duplicate_function_key(node)))
    target_name = passthrough_target(node)
    is_private = node.name.startswith("_") and not node.name.startswith("__") and node.name not in exported_names
    if target_name and is_private:
        passthrough_functions.append((node.name, node.lineno, end_line, target_name))
    if (
        loaded_name_counts.get(node.name, 0) == 1
        and is_private
        and target_name is None
        and not node.decorator_list
        and length <= INLINE_FUNCTION_MAX_LINES
    ):
        inline_functions.append((node.name, node.lineno, end_line, 1))
    if length < LONG_FUNCTION_THRESHOLD:
        return
    candidates.append(
        Candidate(
            id=f"py-extract-function-{rel_path}-{node.lineno}-{node.name}",
            kind="extract_function",
            title=f"Extract logic from long function {node.name}",
            description=(
                f"Function `{node.name}` in {rel_path} spans {length} lines and is a candidate for extraction"
            ),
            language="python",
            scope="local",
            source=["static", "metric"],
            files=[rel_path],
            symbols=[node.name],
            anchorRegions=[region(rel_path, node.lineno, end_line)],
            estimatedBenefit=benefit(
                {
                    "complexityReduction": min(1.0, length / max(len(lines), 1)),
                    "maintainabilityGain": 0.35,
                }
            ),
            estimatedRisk=risk({"semanticRisk": 0.35, "testRisk": 0.25, "conflictRisk": 0.15}),
            estimatedDiff=diff({"filesTouched": 1, "linesAdded": max(3, length // 4), "linesModified": length}),
            confidence=0.72,
            applyModeHint="guarded",
            requiredChecks=["parse", "lint", "typecheck", "unit_test"],
            provenance=Provenance(
                detectors=["python-ast-long-function"],
                evidence=[f"line_span:{length}", f"symbol:{node.name}"],
            ),
        )
    )


def dead_code_candidates(
    rel_path: str,
    tree: ast.Module,
    referenced_names: set[str],
    exported_names: set[str],
) -> list[Candidate]:
    candidates: list[Candidate] = []
    for node in tree.body:
        if isinstance(node, (ast.Assign, ast.AnnAssign)):
            candidate = top_level_unused_assignment_candidate(rel_path, node, referenced_names, exported_names)
            if candidate is not None:
                candidates.append(candidate)
        elif isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)) and node.end_lineno is not None:
            if _is_dead_private_symbol(node.name, referenced_names, exported_names):
                candidates.append(_dead_function_candidate(rel_path, node))
        elif isinstance(node, ast.ClassDef) and node.end_lineno is not None:
            if _is_dead_private_symbol(node.name, referenced_names, exported_names):
                candidates.append(_dead_class_candidate(rel_path, node))
    return candidates


def _is_dead_private_symbol(name: str, referenced_names: set[str], exported_names: set[str]) -> bool:
    return name.startswith("_") and not name.startswith("__") and name not in referenced_names and name not in exported_names


def _dead_function_candidate(rel_path: str, node: ast.FunctionDef | ast.AsyncFunctionDef) -> Candidate:
    end_line = node.end_lineno
    assert end_line is not None
    length = end_line - node.lineno + 1
    return Candidate(
        id=f"py-dead-code-{rel_path}-{node.lineno}-{node.name}",
        kind="dead_code",
        title=f"Remove unused private function {node.name}",
        description=f"Top-level private function `{node.name}` in {rel_path} is not referenced",
        language="python",
        scope="module",
        source=["static"],
        files=[rel_path],
        symbols=[node.name],
        anchorRegions=[region(rel_path, node.lineno, end_line)],
        estimatedBenefit=benefit({"maintainabilityGain": 0.18}),
        estimatedRisk=risk({"semanticRisk": 0.08, "conflictRisk": 0.04}),
        estimatedDiff=diff({"filesTouched": 1, "linesDeleted": length, "linesModified": length}),
        confidence=0.86,
        applyModeHint="auto",
        requiredChecks=["parse", "lint", "typecheck"],
        provenance=Provenance(
            detectors=["python-ast-dead-code"],
            evidence=[f"line_span:{length}", f"symbol:{node.name}"],
        ),
    )


def _dead_class_candidate(rel_path: str, node: ast.ClassDef) -> Candidate:
    end_line = node.end_lineno
    assert end_line is not None
    length = end_line - node.lineno + 1
    return Candidate(
        id=f"py-dead-code-{rel_path}-{node.lineno}-{node.name}",
        kind="dead_code",
        title=f"Remove unused private class {node.name}",
        description=f"Top-level private class `{node.name}` in {rel_path} is not referenced",
        language="python",
        scope="module",
        source=["static"],
        files=[rel_path],
        symbols=[node.name],
        anchorRegions=[region(rel_path, node.lineno, end_line)],
        estimatedBenefit=benefit({"maintainabilityGain": 0.18}),
        estimatedRisk=risk({"semanticRisk": 0.1, "conflictRisk": 0.05}),
        estimatedDiff=diff({"filesTouched": 1, "linesDeleted": length, "linesModified": length}),
        confidence=0.84,
        applyModeHint="auto",
        requiredChecks=["parse", "lint", "typecheck"],
        provenance=Provenance(
            detectors=["python-ast-dead-code"],
            evidence=[f"line_span:{length}", f"symbol:{node.name}"],
        ),
    )
